consensus/view_change.go

```diff
@@ -90,20 +90,111 @@ func (pm *State) GetViewChangeDuraion() time.Duration {
 	return time.Duration(diff * diff * int64(viewChangeDuration))
 }
 
-// GetNextLeaderKey uniquely determine who is the leader for given viewID
-func (consensus *Consensus) GetNextLeaderKey(viewID uint64) *bls.PublicKeyWrapper {
+// fallbackNextViewID return the next view ID and duration when there is an exception
+// to calculate the time-based viewId
+func (consensus *Consensus) fallbackNextViewID(viewID uint64) (uint64, time.Duration) {
+	consensus.getLogger().Error().
+		Uint64("viewID", viewID).
+		Msg("[fallbackNextViewID] use legacy viewID algorithm")
+
+	if viewID < consensus.current.GetCurBlockViewID() {
+		return consensus.current.GetCurBlockViewID() + 1, time.Duration(int64(viewChangeDuration))
+	}
+	diff := int64(viewID + 1 - consensus.current.GetCurBlockViewID())
+	return viewID + 1, time.Duration(diff * diff * int64(viewChangeDuration))
+}
+
+// getNextViewID return the next view ID based on the timestamp
+// The next view ID is calculated based on the difference of validator's timestamp
+// and the block's timestamp. So that it can be deterministic to return the next view ID
+// only based on the blockchain block and the validator's current timestamp.
+// The next view ID is the single factor used to determine
+// the next leader, so it is mod the number of nodes per shard.
+// It returns the next viewID and duration of the view change
+// The view change duration is a fixed duration now to avoid stuck into offline nodes during
+// the view change.
+// viewID is only used as the fallback mechansim to determine the nextViewID
+func (consensus *Consensus) getNextViewID(viewID uint64) (uint64, time.Duration) {
+	// handle corner case at first
+	if consensus.ChainReader == nil {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	curHeader := consensus.ChainReader.CurrentHeader()
+	if curHeader == nil {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	blockTimestamp := curHeader.Time().Int64()
+	curTimestamp := time.Now().Unix()
+
+	// timestamp messed up in current validator node
+	if curTimestamp < blockTimestamp {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	totalNode := consensus.Decider.ParticipantsCount()
+	// diff is at least 1, and it won't exceeded the totalNode
+	diff := uint64(((curTimestamp - blockTimestamp) / viewChangeTimeout) % int64(totalNode))
+	if diff == 0 {
```

Comment:
the case of 0 is normal too. I can be that curtimestamp < blockTimestamp + viewChangeTimeout, or the view change goes round the whole list and comes back to the first leader again. So don't need to do diff = 1 here.

Reply:
make sense. will remove this additional check.